source/volume_of_fluid/assembler.cc

```diff
@@ -0,0 +1,667 @@
+                  .get_function_values(this->get_old_solution(),
+                                       scratch.neighbor_old_values);
+                  scratch.neighbor_finite_element_values[volume_of_fluidN_n]
+                  .get_function_values(this->get_old_solution(),
+                                       scratch.neighbor_i_n_values);
+                  scratch.neighbor_finite_element_values[volume_of_fluidN_d]
+                  .get_function_values(this->get_old_solution(),
+                                       scratch.neighbor_i_d_values);
+                }
+              else
+                {
+                  scratch.neighbor_finite_element_values[solution_field]
+                  .get_function_values(this->get_solution(),
+                                       scratch.neighbor_old_values);
+                  scratch.neighbor_finite_element_values[volume_of_fluidN_n]
+                  .get_function_values(this->get_solution(),
+                                       scratch.neighbor_i_n_values);
+                  scratch.neighbor_finite_element_values[volume_of_fluidN_d]
+                  .get_function_values(this->get_solution(),
+                                       scratch.neighbor_i_d_values);
+                }
+
+              const double neighbor_vol = neighbor->measure();
+              const double neighbor_volume_of_fluid = scratch.neighbor_old_values[0];
+              const Tensor<1, dim, double> neighbor_i_normal = scratch.neighbor_i_n_values[0];
+              const double neighbor_i_d = scratch.neighbor_i_d_values[0];
+
+              double face_flux = 0;
+              double face_ls_d = 0;
+              double face_ls_time_grad = 0;
+              double n_face_ls_d = 0;
+              double n_face_ls_time_grad =0;
+
+              // Using VolumeOfFluid so need to accumulate flux through face
+              for (unsigned int q=0; q<n_f_q_points; ++q)
+                {
+
+                  Tensor<1,dim> current_u = scratch.face_current_velocity_values[q];
+
+                  //If old velocity available average to half timestep
+                  if (old_velocity_avail)
+                    current_u += 0.5*(scratch.face_old_velocity_values[q] -
+                                      scratch.face_current_velocity_values[q]);
+
+                  //Subtract off the mesh velocity for ALE corrections if necessary
+                  if (this->get_parameters().free_surface_enabled)
+                    current_u -= scratch.face_mesh_velocity_values[q];
+
+                  face_flux += this->get_timestep() *
+                               current_u *
+                               scratch.face_finite_element_values.normal_vector(q) *
+                               scratch.face_finite_element_values.JxW(q);
+
+                }
+
+              // Due to inability to reference this cell's values at the interface,
+              // need to do explicit calculation
```

Comment:
Lets talk about this in more detail. What can you not access, and what are you calculating here?